### ISSUE TITLE
keras metrics will not accept complex data

### ISSUE THREAD

**Issue by isaacgerg** [keras-team/keras#16172]
TF 2.4.1 - 2.7

When using a metric that accepts complex numbers, tf keras fit ends up discarding the imaginary part of the complex number and then casts the array to real when inputting y_pred and y_true to a metric function.

It works correctly when used as a loss function.  The casting only occurs when the loss function is used as a metric.

The error is because of casting done in metrics.py on line 609.  See the snippet below from this file.

```
  def update_state(self, y_true, y_pred, sample_weight=None):
    """Accumulates metric statistics.

    `y_true` and `y_pred` should have the same shape.

    Args:
      y_true: Ground truth values. shape = `[batch_size, d0, .. dN]`.
      y_pred: The predicted values. shape = `[batch_size, d0, .. dN]`.
      sample_weight: Optional `sample_weight` acts as a
        coefficient for the metric. If a scalar is provided, then the metric is
        simply scaled by the given value. If `sample_weight` is a tensor of size
        `[batch_size]`, then the metric for each sample of the batch is rescaled
        by the corresponding element in the `sample_weight` vector. If the shape
        of `sample_weight` is `[batch_size, d0, .. dN-1]` (or can be broadcasted
        to this shape), then each metric element of `y_pred` is scaled by the
        corresponding value of `sample_weight`. (Note on `dN-1`: all metric
        functions reduce by 1 dimension, usually the last axis (-1)).

    Returns:
      Update op.
    """
    y_true = math_ops.cast(y_true, self._dtype)  # THIS IS WRONG!
    y_pred = math_ops.cast(y_pred, self._dtype)
    [y_true, y_pred], sample_weight = \
        metrics_utils.ragged_assert_compatible_and_get_flat_values(
            [y_true, y_pred], sample_weight)
    y_pred, y_true = losses_utils.squeeze_or_expand_dimensions(
        y_pred, y_true)

    ag_fn = autograph.tf_convert(self._fn, ag_ctx.control_status_ctx())
    matches = ag_fn(y_true, y_pred, **self._fn_kwargs)
    return super(MeanMetricWrapper, self).update_state(
        matches, sample_weight=sample_weight)
```

Setting a break point at the start of this function, we see the following vars in debug:
y_true
<tf.Tensor 'IteratorGetNext:3' shape=(None, 256, 256, 1) dtype=complex64>
self._dtype
'float32'

y_true should remain of type complex64 but instead is casted to float32 by ignoring the imaginary part and keeping the real part.

A correct solution will make an additional check for complex data in addition to the system floatx() type and cast accordingly (i.e. if floatx() is float32, complex128 becomes complex64; float64 goes to float32).

 I searched for this pattern,  tf.cast(y_true, self._dtype), and it occurs quite a bit throughout this file.  From what I can tell, all cases are incorrect and will cast complex data to float data and discard the imaginary part.  I also confirmed this bug exists in tf 2.7, reference https://github.com/keras-team/keras/blob/v2.7.0/keras/metrics.py#L1096-L1141

**Comment by haifeng-jin** [comment]
Summary: The main problem is metrics would cast `y` to `self._dtype`, which is the `dtype` of the metric value, which may not be the correct `dtype` for `y`.

**Comment by fchollet** [comment]
The cast is likely useful in many cases, when users are passing mismatched but resolvable (castable) dtypes. But in this case it produces the wrong result. We could restrict the casting to only convert float/int to float, for instance.

**Comment by isaacgerg** [comment]
My opinion is to do no casting whatsoever and leave it up to the programmer to ensure they are using the methods according to the documentation. At most, I would put exceptions around to prevent any kind transform on complex numbers unless explicitly stated by the programmer.
